Documentation/mscorlib.md

```diff
@@ -0,0 +1,357 @@
+
+[ecalllist]: https://github.com/dotnet/coreclr/blob/master/src/vm/ecalllist.h
+
+class FooNative
+    {
+    public:
+         // All QCalls should be static and should be tagged with QCALLTYPE
+         static
+         BOOL QCALLTYPE Bar(int flags, LPCWSTR wszString, QCall::StringHandleOnStack retString);
+    };
+    
+    BOOL QCALLTYPE FooNative::Bar(int flags, LPCWSTR wszString, QCall::StringHandleOnStack retString)
+    {
+         // All QCalls should have QCALL_CONTRACT.
+         // It is alias for THROWS; GC_TRIGGERS; MODE_PREEMPTIVE; SO_TOLERANT.
+         QCALL_CONTRACT;
+
+         // Optionally, use QCALL_CHECK instead and the expanded form of the contract
+         // if you want to specify preconditions:
+         // CONTRACTL { 
+         //     QCALL_CHECK; 
+         //     PRECONDITION(wszString != NULL);
+         // } CONTRACTL_END;
+       
+         // The only line between QCALL_CONTRACT and BEGIN_QCALL
+         // should be the return value declaration if there is one.
+         BOOL retVal = FALSE;
+
+         // The body has to be enclosed in BEGIN_QCALL/END_QCALL macro. It is necessary 
+         // to make the exception handling work.
+         BEGIN_QCALL;
+
+         // Validate arguments if necessary and throw exceptions.
+         // There is no convention currently on whether the argument validation should be 
+         // done in managed or unmanaged code.
+         if (flags != 0)
+             COMPlusThrow(kArgumentException, L"InvalidFlags");
+
+         // No need to worry about GC moving strings passed into QCall.
+         // Marshalling pins them for us.
+         printf("%S", wszString);
+
+         // This is most the efficient way to return strings back 
+         // to managed code. No need to use StringBuilder.
+         retString.Set(L"Hello");
+
+         // You can not return from inside of BEGIN_QCALL/END_QCALL. 
+         // The return value has to be passed out in helper variable.
+         retVal = TRUE;
+
+         END_QCALL;
+
+         return retVal;
+     }
+
+## FCall Functional Behavior
+
+FCalls allow more flexibility in terms of passing object references around, with a higher code complexity and more opportunities to hang yourself. Additionally, FCall methods must either erect a helper method frame along their common code paths, or for any FCall of non-trivial length, explicitly poll for whether a garbage collection must occur. Failing to do so will lead to starvation issues if managed code repeatedly calls the FCall method in a tight loop, because FCalls execute while the thread only allows the GC to run in a cooperative manner.
+
+FCalls require a lot of glue, too much to describe here. Look at [fcall.h][fcall] for details.
+
+[fcall]: https://github.com/dotnet/coreclr/blob/master/src/vm/fcall.h
+
+### GC Holes, FCall, and QCall
+
+TODO: Publish CLR-Coding-Guide to GitHub
```

Comment:
That document is almost done, too, so it is a reminder to update the link in the right place in the doc.